geyser-plugin-manager/src/geyser_plugin_manager.rs

```diff
@@ -29,21 +38,115 @@ impl GeyserPluginManager {
         libpath: &str,
         config_file: &str,
     ) -> Result<(), Box<dyn Error>> {
-        type PluginConstructor = unsafe fn() -> *mut dyn GeyserPlugin;
+        // Load the dynamic library and reconstruct the Box<dyn GeyserPlugin>
         let lib = Library::new(libpath)?;
         let constructor: Symbol<PluginConstructor> = lib.get(b"_create_plugin")?;
         let plugin_raw = constructor();
         let mut plugin = Box::from_raw(plugin_raw);
+
+        // Run on_load to do necessary initializations
         plugin.on_load(config_file)?;
-        self.plugins.push(plugin);
-        self.libs.push(lib);
+
+        // Wrap in Arc<RwLock<_>>
+        let plugin = Arc::new(RwLock::new(plugin));
+        let lib = Arc::new(RwLock::new(lib));
+
+        self.plugins.push(Arc::clone(&plugin));
+        self.libs.push(Arc::clone(&lib));
+
+        let owned_path = std::path::PathBuf::from(libpath);
+        let owned_config = config_file.to_owned();
+        let handle = std::thread::spawn(move || {
+            let mut refresh_path = owned_path.clone();
+            refresh_path.set_extension("");
+            let mut done_path = refresh_path.clone();
+            refresh_path.set_file_name(format!(
+                "refresh_{}",
+                refresh_path
+                    .file_name()
+                    .map(std::ffi::OsStr::to_str)
+                    .expect("valid path with valid filename was given if library loaded")
+                    .expect("valid path with valid filename was given if library loaded")
+            ));
+            done_path.set_file_name(format!(
+                "result_{}",
+                refresh_path
+                    .file_name()
+                    .map(std::ffi::OsStr::to_str)
+                    .expect("valid path with valid filename was given if library loaded")
+                    .expect("valid path with valid filename was given if library loaded")
+            ));
+
+            #[allow(unused_must_use)] // result io
+            loop {
+                // Every 10 seconds..
+                std::thread::park_timeout(Duration::from_secs(10));
+
+                info!("checking for new plugin at {}", refresh_path.display());
+
+                // If refresh path exists, handle refresh request
+                if refresh_path.exists() {
+                    info!("detected refresh_path {}", refresh_path.display());
```

Comment:
use a channel instead

Reply:
instead of a park_timeout? That's a full-on block unless it's async